harness/tests/experiment/pytorch/test_pytorch_context.py

```diff
@@ -0,0 +1,50 @@
+import unittest
+
+import torch
+
+from determined import pytorch
+from determined.common.check import CheckFailedError
+from determined.errors import InternalException
+from tests.experiment.fixtures import pytorch_onevar_model
+
+
+class TestPyTorchContext(unittest.TestCase):
+    def setUp(self) -> None:
+        self.config = {"hyperparameters": {"global_batch_size": 4, "dataloader_type": "determined"}}
+        self.context: pytorch.PyTorchTrialContext = pytorch.PyTorchTrialContext.from_config(
+            self.config
+        )
+        assert isinstance(self.context, pytorch.PyTorchTrialContext)
+
+        trial = pytorch_onevar_model.OneVarTrial(self.context)
+
+        train_ds = trial.build_training_data_loader()
+        for epoch_idx in range(3):
+            for batch_idx, batch in enumerate(train_ds):
+                metrics = trial.train_batch(batch, epoch_idx, batch_idx)
+                # Verify the training is correct.
+                pytorch_onevar_model.OneVarTrial.check_batch_metrics(metrics, batch_idx)
+
+        eval_ds = trial.build_validation_data_loader()
+        for batch in eval_ds:
+            metrics = trial.evaluate_batch(batch)
+
+    def test_average_gradients(self) -> None:
+        self.assertRaises(CheckFailedError, self.context._average_gradients, None, 0)
+        self.assertIsNone(self.context._average_gradients(None, 1))
+
+    def test_training_not_started(self) -> None:
+        self.assertRaises(InternalException, self.context.is_epoch_start)
+        self.assertRaises(InternalException, self.context.is_epoch_end)
+        self.assertRaises(InternalException, self.context.current_train_batch)
+        self.assertRaises(InternalException, self.context.current_train_epoch)
+        self.context.env.managed_training = True
+        self.assertRaises(InternalException, self.context._should_communicate_and_update)
+
+    def test_wrap_scalar(self) -> None:
+        scaler = 1
```

Comment:
I'd recommend using a scaler object instead of an int here.

Reply:
A scaler object would be like [this one](https://pytorch.org/docs/stable/amp.html#torch.cuda.amp.GradScaler).

Reply:
This makes sense, but I note that `wrap_scaler` has the type hinting `scaler: Any`, which is very general. Anyway, my using `1` is wrong and was due to my misunderstanding it as a "`scalar`".

Reply:
The reason it is an `Any` is because it's tricky to have it be more precise, since the actual object type doesn't exist in some of the earlier forms of pytorch that we support.